march_hardware_interface/src/march_hardware_interface.cpp

```diff
@@ -0,0 +1,119 @@
+// Copyright 2019 Project March.
+
+#include <sstream>
+#include <march_hardware_interface/march_hardware_interface.h>
+#include <joint_limits_interface/joint_limits_interface.h>
+#include <joint_limits_interface/joint_limits.h>
+#include <joint_limits_interface/joint_limits_urdf.h>
+#include <joint_limits_interface/joint_limits_rosparam.h>
+#include <march_hardware/March4.h>
+
+using joint_limits_interface::JointLimits;
+using joint_limits_interface::SoftJointLimits;
+using joint_limits_interface::PositionJointSoftLimitsHandle;
+using joint_limits_interface::PositionJointSoftLimitsInterface;
+
+namespace march_hardware_interface
+{
+MarchHardwareInterface::MarchHardwareInterface(ros::NodeHandle& nh) : nh_(nh)
+{
+  init();
+  controller_manager_.reset(new controller_manager::ControllerManager(this, nh_));
+  nh_.param("/march/hardware_interface/loop_hz", loop_hz_, 0.1);
+  ros::Duration update_freq = ros::Duration(1.0 / loop_hz_);
+  non_realtime_loop_ = nh_.createTimer(update_freq, &MarchHardwareInterface::update, this);
+}
+
+MarchHardwareInterface::~MarchHardwareInterface()
+{
+}
+
+void MarchHardwareInterface::init()
+{
+  this->march = march4cpp::MARCH4();
+  this->march.startEtherCAT();
+
+  if (!this->march.isEthercatOperational())
+  {
+    ROS_FATAL("EtherCAT is not operational");
+    exit(0);
+  }
+
+  // Get joint names
+  nh_.getParam("/march/hardware_interface/joints", joint_names_);
+  num_joints_ = joint_names_.size();
+
+  // Resize vectors
+  joint_position_.resize(num_joints_);
+  joint_velocity_.resize(num_joints_);
+  joint_effort_.resize(num_joints_);
+  joint_position_command_.resize(num_joints_);
+  joint_velocity_command_.resize(num_joints_);
+  joint_effort_command_.resize(num_joints_);
+
+  // Initialize Controller
+  for (int i = 0; i < num_joints_; ++i)
+  {
+    march4cpp::Joint joint = march.getJoint(joint_names_[i]);
+    // Create joint state interface
+    JointStateHandle jointStateHandle(joint.getName(), &joint_position_[i], &joint_velocity_[i], &joint_effort_[i]);
+    joint_state_interface_.registerHandle(jointStateHandle);
+
+    // Create position joint interface
+    JointHandle jointPositionHandle(jointStateHandle, &joint_position_command_[i]);
+    JointLimits limits;
+    SoftJointLimits softLimits;
+
+    getJointLimits(joint.getName(), nh_, limits);
+    getSoftJointLimits(joint.getName(), nh_, softLimits);
+
+    PositionJointSoftLimitsHandle jointLimitsHandle(jointPositionHandle, limits, softLimits);
+    positionJointSoftLimitsInterface.registerHandle(jointLimitsHandle);
+    position_joint_interface_.registerHandle(jointPositionHandle);
```

Comment:
Is the joint limitshandle a separate interface or is it part of the position joint interface ? If not, maybe consider adding a comment above it, as with the other interfaces

Reply:
Seperate. I'm not sure what kind of comment you'd want here

Reply:
```suggestion
    // Create joint limit interface
```
This is more inline with the comments that were placed before